### PR TITLE
make writebuffers cancelable

### LINKED ISSUE #64451: `test_query_is_canceled_with_inf_retries` flaks due to miss implementation for cancelation logic
Some time `test_checking_s3_blobs_paranoid/test.py::test_query_is_canceled_with_inf_retries` flaks.
Lile this:
https://s3.amazonaws.com/clickhouse-test-reports/0/ae4d3f97ae6345b81efba270602522755d575940/integration_tests__aarch64__%5B5_6%5D.html

What test do:
- set up a faulty disk which closes incoming confections as `Connection refused` 
- fire `INSERT` request which write data to that disk with infinite retries
- after some time test fires `KILL QUERY` to make sure that such `INSERT` requests are killable

The test fails because `KILL QUERY` requests fails with exceptions from `INSERT` request. That is not expected. `KILL QUERY` is supposed just kill other query and does not inherit its exceptions.

That happens on a race condition.
If `INSERT` had a chance to met the upload limits and stoped uploading on `WriteBufferFromS3::nextImpl()::writeMultipartUpload()::writePart()::task_tracker->add()::waitTilInflightShrink().`
Than `INSERT` observes exception when it is killed. Because infinite retries stop when they see the cancelation flag and as a result the last exception is throw. That is correct.
What is not correct that `KILL QUERY` calls `process_list.sendCancelToQuery():: elem->cancelQuery(kill)::executors_snapshot->cancel()::graph->cancel()::processor->cancel()::onCancel()` and method `IProcessor::onCancel()` has inapprpriate implementation.

For `StorageS3Sink::onCancel()` `finalize()` is called which rethrow all occured exception to the `KILL QUERY`.
For `MergeTreeSink::onCancel()` nothing is called at all which is kinda OK, but it produces a lot of warning/errors logs because we destroy unfinished write buffer after that without exception context. It would be more neat if explicit cancel is called for write buffer here.

### PR DESCRIPTION
**Pull request by CheSema**
fixes https://github.com/ClickHouse/ClickHouse/issues/64451

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix bug with cancelation buffer when canceling a query.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>CI Settings</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step
- [ ]  Allow: Integration Tests
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Unit tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All with aarch64
- [ ]  Allow: All with ASAN
- [ ]  Allow: All with TSAN
- [ ]  Allow: All with Analyzer
- [ ]  Allow: All with Azure
- [ ]  Allow: Add your option here
---
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN
- [ ]  Exclude: All with MSAN
- [ ]  Exclude: All with UBSAN
- [ ]  Exclude: All with Coverage
- [ ]  Exclude: All with Aarch64
---
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)
- [ ]  allow: batch 1 for multi-batch jobs
- [ ]  allow: batch 2
- [ ]  allow: batch 3
- [ ]  allow: batch 4, 5 and 6
</details>

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
*This is an automated comment for commit eb35d17e72b531ac56d552869716e825923a0ef6 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64478/eb35d17e72b531ac56d552869716e825923a0ef6/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by antonio2368 — APPROVED** [review]
👍

File: src/IO/WriteBuffer.h [hunk 1]
```diff
@@ -150,7 +172,12 @@ class WriteBuffer : public BufferBase
         next();
     }
 
+    virtual void cancelImpl()
```
Comment (antonio2368) [inline src/IO/WriteBuffer.h:175]: actually why not do the same in read buffers?

Comment (CheSema) [reply src/IO/WriteBuffer.h:175]: Read buffers are suppose to be cont methods related to the storage. It is just OK to destroy them at any point.

WriteBuffers should be destroyed only after `finalize` call. It is a source of all difficulties. But that `finalize` call is essential.

**Comment by CheSema** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/64478/517c1138113220e5a4318c07806759911896234d/stress_test__debug_/fatal_messages.txt

this is related to the changes

**Comment by CheSema** [comment]
@antonio2368 could I ask you to check this once again?
I think I passed the CI. The last CI has a crash which turned out is not raleted to my changes.

File: src/Client/Connection.cpp [hunk 2]
```diff
@@ -37,6 +38,7 @@
 #include <Common/FailPoint.h>
 
 #include <Common/config_version.h>
+#include "Core/Types.h"
```
Comment (antonio2368) [inline src/Client/Connection.cpp:41]:
```suggestion
#include <Core/Types.h>
```

**Review by antonio2368 — APPROVED** [review]

**Comment by CheSema** [comment]
01926_order_by_desc_limit -- https://github.com/ClickHouse/ClickHouse/issues/65472
02922_deduplication_with_zero_copy -- https://github.com/ClickHouse/ClickHouse/issues/65486
test_when_s3_broken_pipe_at_upload_is_retried -- related to https://github.com/ClickHouse/ClickHouse/issues/65472
00988_parallel_parts_removal -- checking

**Comment by CheSema** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/64478/19bafb5063ece9d71f44f248b35087800c4f302c/stress_test__ubsan_.html

```
 [ 4096 ] {a7c87804-bf65-4264-863f-d34164941762} <Fatal> : Logical error: \'false && "WriteBuffer is not finalized in destructor."\'.
```
related to the issue

**Comment by CheSema** [comment]
02956_rocksdb_bulk_sink flaks in master.

**Pull request merged**